File: fszmq.props/fszmq.props.vb/Observations.vb
Imports FsCheck.PropertyExtensions
Imports CheckProperty = FsCheck.NUnit.PropertyAttribute

Namespace fszmq.props

  <Extension()>
  Public Module Classifiers
  
    <Extension()>
    Function Within (ByVal item  As Integer, 
                     ByVal min   As Integer, 
                     ByVal max   As Integer) As Boolean
      Return (item >= min) AndAlso (item <= max)
    End Function
    
    <Extension()>
    Function IsTiny(ByVal input As Mod4Binary) As Boolean
      Return input.Data.Length.Within(Integer.MinValue,4)
    End Function
    
    <Extension()>
    Function IsSmall(ByVal input As Mod4Binary) As Boolean
      Return input.Data.Length.Within(8,20)
    End Function
    
    <Extension()>
    Function IsMedium(ByVal input As Mod4Binary) As Boolean
      Return input.Data.Length.Within(24,60)
    End Function

    <Extension()>
    Function IsLarge(ByVal input As Mod4Binary) As Boolean
      Return input.Data.Length.Within(64,Integer.MaxValue)
    End Function

  End Module

  <TestFixture()> _
  Public Class Observations

    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Public Function EncodeDecodeAreDuals_Trivial (ByVal input As Mod4Binary) As [Property]
      Dim processed = Z85.Decode(Z85.Encode(input.Data))
      ' identify a special case
      Return processed.SequenceEqual(input.Data) _
                      .Trivial(input.Data.Any(Function (b) b = 70))
    End Function

    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Public Function EncodeDecodeAreDuals_Classified (ByVal input As Mod4Binary) As [Property]
      Dim processed = Z85.Decode(Z85.Encode(input.Data))
      ' bucket by common sizes
      Return processed.SequenceEqual(input.Data)                              _
                      .Classify(input.IsTiny  (), "Tiny (0 .. 4 bytes)"    )  _
                      .Classify(input.IsSmall (), "Small (8 .. 20 bytes)"  )  _
                      .Classify(input.IsMedium(), "Medium (24 .. 60 bytes)")  _
                      .Classify(input.IsLarge (), "Large (64 .. ∞ bytes)"  )
    End Function

    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Public Function EncodeDecodeAreDuals_Collected (ByVal input As Mod4Binary) As [Property]
      Dim processed = Z85.Decode(Z85.Encode(input.Data))
      ' gather individual sizes
      Return processed.SequenceEqual(input.Data) _
                      .Collect(input.Data.Length)
    End Function

    <CheckProperty(Arbitrary := { GetType(Generators) })>
    Public Function EncodeDecodeAreDuals_Combined (ByVal input As Mod4Binary) As [Property]
      Dim processed = Z85.Decode(Z85.Encode(input.Data))
      ' identify a special case, bucket by common sizes, and gather individual sizes
      Return processed.SequenceEqual(input.Data)                              _
                      .Trivial  (input.Data.Any(Function (b) b = 70))         _
                      .Classify (input.IsTiny  (), "Tiny (0 .. 4 bytes)"    ) _
                      .Classify (input.IsSmall (), "Small (8 .. 20 bytes)"  ) _
                      .Classify (input.IsMedium(), "Medium (24 .. 60 bytes)") _
                      .Classify (input.IsLarge (), "Large (64 .. ∞ bytes)"  ) _
                      .Collect  (input.Data.Length)  
    End Function

  End Class

End Namespace

File: fszmq.props/fszmq.props.cs/Observations.cs
namespace fszmq.props
{
  using NUnit.Framework;
  using FsCheck;
  using System;
  using System.Linq;
  using System.Runtime.CompilerServices;

  using CheckProperty = FsCheck.NUnit.PropertyAttribute;

  using fszmq;

  public static class Classifiers
  {
    public static Boolean Within (this Int32 item, Int32 min, Int32 max)
    {
      return (item >= min) && (item <= max);
    }

    public static Boolean IsTiny(this Mod4Binary input)
    {
      return input.Data.Length.Within(Int32.MinValue, 4);
    }

    public static Boolean IsSmall(this Mod4Binary input)
    {
      return input.Data.Length.Within(8, 20);
    }

    public static Boolean IsMedium(this Mod4Binary input) 
    {
      return input.Data.Length.Within(24, 60);
    }

    public static Boolean IsLarge(this Mod4Binary input)
    {
      return input.Data.Length.Within(64, Int32.MaxValue);
    }
  }

  [TestFixture]
  public class Observations
  {
    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property EncodeDecodeAreDuals_Trivial (Mod4Binary input)
    {
      var processed = Z85.Decode(Z85.Encode(input.Data));
      return processed.SequenceEqual(input.Data)
                      // identify a special case
                      .Trivial(input.Data.Any (b => b == 70));
    }

    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property EncodeDecodeAreDuals_Classified (Mod4Binary input)
    {
      var processed = Z85.Decode(Z85.Encode(input.Data));
      return processed.SequenceEqual(input.Data)
                      // bucket by common sizes
                      .Classify(input.IsLarge (),"Large (64 .. ∞ bytes)"  )
                      .Classify(input.IsMedium(),"Medium (24 .. 60 bytes)")
                      .Classify(input.IsSmall (),"Small (8 .. 20 bytes)"  )
                      .Classify(input.IsTiny  (),"Tiny (0 .. 4 bytes)"    );
    }

    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property EncodeDecodeAreDuals_Collected (Mod4Binary input)
    {
      var processed = Z85.Decode(Z85.Encode(input.Data));
      return processed.SequenceEqual(input.Data)
                      // gather individual sizes
                      .Collect(input.Data.Length);
    }

    [CheckProperty(Arbitrary = new []{ typeof(Generators) })]
    public Property EncodeDecodeAreDuals_Combined (Mod4Binary input)
    {
      var processed = Z85.Decode(Z85.Encode(input.Data));
      return processed.SequenceEqual(input.Data)
                      // identify a special case
                      .Trivial(input.Data.Any (b => b == 70))
                      // bucket by common sizes
                      .Classify(input.IsLarge (),"Large (64 .. ∞ bytes)"  )
                      .Classify(input.IsMedium(),"Medium (24 .. 60 bytes)")
                      .Classify(input.IsSmall (),"Small (8 .. 20 bytes)"  )
                      .Classify(input.IsTiny  (),"Tiny (0 .. 4 bytes)"    )
                      // gather individual sizes
                      .Collect(input.Data.Length);
    }
  }
}
